src/Uno.Extensions.Reactive/Collections/Facades/Map/MapReadOnlyList.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Uno.Extensions.Conversion;
+
+namespace Uno.Extensions.Collections.Facades.Map;
+
+/// <summary>
+/// A facade over a readonly list which ensure the dynamic projection of items using an <see cref="IConverter{TFrom,TTo}"/>.
+/// </summary>
+/// <typeparam name="TFrom"></typeparam>
+/// <typeparam name="TTo"></typeparam>
+internal class MapReadOnlyList<TFrom, TTo> : IList, IList<TTo>, IReadOnlyList<TTo>
+{
+	private readonly IConverter<TFrom, TTo> _converter;
+	private readonly IList _source;
+
+	/// <summary>
+	/// Creates a facade over a readonly list which ensure conversion of items using the provided converter.
+	/// </summary>
+	/// <param name="converter"></param>
+	/// <param name="source">Must be read only itself</param>
+	public MapReadOnlyList(IConverter<TFrom, TTo> converter, IList source)
+	{
+		_converter = converter;
+		_source = source;
+
+		if (!source.IsReadOnly)
+		{
+			throw new InvalidOperationException("MapReadOnlyList is only a facade over a list which should be read-only itself.");
```

Comment:
Can the message be more actionable?

Reply:
Changed to an `ArgumentException` to be more explicit about the fact that it's the `source` that is invalid